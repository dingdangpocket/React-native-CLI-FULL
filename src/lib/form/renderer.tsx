import { createElement, FC, memo, ReactNode } from 'react';
import { Text, View } from 'react-native';
import { FormTextInput, FormTextInputProps } from './elements/text';
import { FormData, FormElement } from './types';

interface Props<T extends FormData> {
  element: FormElement<T>;
  value: any;
  onChange: (values: Partial<T>) => void;
}

const TextContent = memo(({ text }: { text?: ReactNode }) => {
  if (!text) return null;
  if (
    typeof text === 'string' ||
    typeof text === 'number' ||
    typeof text === 'boolean'
  ) {
    return <Text>{String(text)}</Text>;
  }
  return <View>{text}</View>;
});

const FormElementLabel = memo(({ label }: { label?: ReactNode }) => {
  return <TextContent text={label} />;
});

const FormHelpText = memo(({ text }: { text?: ReactNode }) => {
  return <TextContent text={text} />;
});

export const FormElementRenderer = memo(
  <T extends FormData>({ element, value, onChange }: Props<T>) => {
    const { label, labelPosition, helpText, ...props } = element;
    const orientation = label && labelPosition === 'left' ? 'h' : 'v';

    return (
      <View
        css={`
          flex-direction: ${orientation === 'h' ? 'row' : 'column'};
          flex-wrap: nowrap;
          justify-content: ${orientation === 'h'
            ? 'space-between'
            : 'flex-start'};
          align-items: ${orientation === 'h' ? 'center' : 'flex-start'};
        `}
      >
        <FormElementLabel label={label} />
        <View>
          {(() => {
            switch (element.type) {
              case 'text': {
                return createElement(
                  FormTextInput as unknown as FC<FormTextInputProps<T>>,
                  {
                    ...props,
                    value,
                    onChange,
                  },
                );
              }
              default:
                return null;
            }
          })()}
        </View>
        <FormHelpText text={helpText} />
      </View>
    );
  },
);
