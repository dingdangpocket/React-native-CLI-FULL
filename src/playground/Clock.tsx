export const Clock = () => {
  return <div>{Date.now()}</div>;
};
