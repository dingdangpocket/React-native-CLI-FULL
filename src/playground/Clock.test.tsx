import renderer from 'react-test-renderer';
import { Clock } from './Clock';

// Mock Date.now function to make it stable across tests.
Date.now = jest.fn(() => 1482363367071);

test('clock renders correctly', () => {
  const tree = renderer.create(<Clock />).toJSON();
  expect(tree).toMatchSnapshot();
});
