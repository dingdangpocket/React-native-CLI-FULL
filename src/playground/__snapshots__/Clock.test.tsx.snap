// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clock renders correctly 1`] = `
<div>
  1482363367071
</div>
`;
